src/array.rs

```diff
@@ -22,6 +23,142 @@ pub type ReadOnly<T, M> = Generic<T, M, marker::ReadOnly>;
+impl<'a, T, M, A> BoundGeneric<'a, T, M, A>
+where
+    M: Mapper,
+    A: Readable + Writable,
+{
+    /// Updates a value that the accessor points to by reading it, modifying it, and writing it.
+    ///
+    /// Note that some MMIO regions (e.g. the Command Ring Pointer field of the Command
+    /// Ring Control Register of the xHCI) may return 0 regardless of the actual values of the
+    /// fields. For these regions, this operation should be called only once.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if `i >= self.len()`.
+    pub fn update_volatile<U>(&mut self, f: U)
+    where
+        U: FnOnce(&mut T),
+    {
+        self.a.update_volatile(f);
+    }
+
+    /// Alias of [`BoundGeneric::update_volatile`].
+    #[deprecated(since = "0.3.1", note = "use `update_volatile`")]
+    pub fn update<U>(&mut self, f: U)
+    where
+        U: FnOnce(&mut T),
+    {
+        self.update_volatile(f);
+    }
+}
+
+/// A derive macro which converts a field struct type into a struct of accessors with same field names.
+/// For a field struct type `T`, add `#[derive(BoundSetGenericOf)]` before the struct definition to derive
+/// accessor struct type `BoundSetGenericOfT<'a, M, A>`. See [`BoundSetGeneric`] for details.
+pub use accessor_macros::BoundSetGenericOf;
+
+/// Combined with proc-macro [`BoundSetGenericOf`], this trait converts array accessors of field struct types into a struct of accessors with same field names.
+///
+/// This trait is intended to be implemented automatically by [`BoundSetGenericOf`] macro expansion. Users should not implement this manually.
+///
+/// # Examples
+///
+/// ```no_run
+/// use accessor::mapper::Identity;
+/// use accessor::array::{BoundSetGeneric, BoundSetGenericOf};
+///
+/// #[repr(C)]
+/// #[derive(Clone, Copy, BoundSetGenericOf)]
+/// struct Foo {
+///     x: u32,
+///     y: u32,
+/// }
+///
+/// // The above derivation creates a struct-of-accessor type called `BoundSetGenericOfFoo` which is roughly equivalent to:
+/// // ```
+/// // struct BoundSetGenericOfFoo {
+/// //     x: accessor::single::ReadWrite::<u32, M>,
+/// //     y: accessor::single::ReadWrite::<u32, M>,
+/// // }
+/// // ```
+/// // The derivation also implements `BoundSetGeneric<Foo, M, A>` so that an `accessor::array::ReadWrite::<Foo, M>` instance
+/// // can be indexed into a `BoundSetGenericOfFoo` item, which has a lifetime bound to the base array accessor.
+///
+/// let mut a = unsafe { accessor::array::ReadWrite::<Foo, M>::new(0x1000, 10, Identity) };
+///
+/// // read `x` field of 0th element of the array.
+/// let x = a.set_at(0).x.read_volatile();
+///
+/// // write 5 as the `y` field of 2nd element of the array.
+/// a.set_at(2).y.write_volatile(5);
+///
+/// ```
+///
+pub trait BoundSetGeneric<T, M, A>
+where
+    M: Mapper,
+    A: AccessorTypeSpecifier + 'static,
+{
+    type BoundSetGenericType<'a>
+    where
+        Self: 'a;
+
+    fn set_at(&self, i: usize) -> Self::BoundSetGenericType<'_>;
+}
+
```

Comment:
This is an example of
* how can we apply `BoundSetGenericOf` derivation to a struct, and
* how can we spawn a struct of accessors by indexing, out from an array of such struct, invoking `.set_at(i)` method in `accessor::array::BoundSetGeneric` trait.

You can also refer to https://github.com/paulsohn/xhci/commit/a2cf6fbd52bfcf596814e79d64f1f619bf009984 and https://github.com/rust-osdev/xhci/issues/158 (related issue)